### PR TITLE
add mergeAST from GraphiQL, refactor using visit() [WIP]

### PR DESCRIPTION
**Pull request by acao**
Accomplishes https://github.com/graphql/graphiql/issues/836, originally introduced via https://github.com/graphql/graphiql/pull/762 earlier this year, `mergeAST` is a handy utility for in-lining fragments for complex queries.

As per @IvanGoncharov's suggestion, I refactored it to use `language/visitor`'s `visit` function.  Good thing I'd been working with AST and visitors on another project using antlr4 earlier this year, or I would've been so lost here, haha. That said, GraphQL's `visit()` is so powerful, I'm blown away at how simple this was.

Thank you @RamonSaboya for the inspiration, and some solid unit tests, and an excellent idea all around.

I'm wondering if a `Set` would be a more appropriate data structure here?

**This is a WIP**, just learning about `visit()` and this is just a learning excercise for me. 
Not a huge priority to any part of the ecosystem currently, and a lot of important edge cases are still broken here.

### CODE REVIEW

**Comment by acao** [comment]
I found a failing case that wasn't in the original unit tests:
```graphql
      query Test1 {
        ...Fragment4
        ...Fragment5
      }
  
      fragment Fragment5 on Test1 {
        ...Fragment4
      }

      fragment Fragment4 on Test1 {
        id
      }
```
does not give me

```graphql
 query Test1 {
        ...on Test1 {
          id
        }
        ...on Test1 {
          ...on Test1 {
            id
          }
        }
      }
```
instead, I get:
```graphql
 query Test1 {
        ...on Test1 {
          id
        }
        ...on Test1 {
          id
        }
      }
```
this works in the original! hmmm 🤔

File: src/utilities/mergeAST.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+export function mergeAST(ast: DocumentNode): DocumentNode {
+  const fragmentDefinitions: { [name: string]: FragmentDefinitionNode } = {};
+  const inlinedFragments: { [name: string]: FragmentSpreadNode } = {};
+
+  visit(ast, {
+    FragmentDefinition(node) {
+      // Collect the existing FragmentDefinition nodes.
+      fragmentDefinitions[node.name.value] = node;
+    },
+  });
```
Comment (IvanGoncharov) [inline src/utilities/mergeAST.js]: You don't need `visit` here, simple `for of` over `definitions` would be enough.

File: src/utilities/__tests__/mergeAST-test.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import { mergeAST } from '../mergeAST';
+
+import { fixtures } from './mergeAST-fixture';
+
+describe('MergeAST', () => {
+  fixtures.forEach(fixture => {
+    it(fixture.desc, () => {
+      const result = print(mergeAST(parse(fixture.query))).replace(/\s/g, '');
+      expect(result).to.equal(fixture.mergedQuery.replace(/\s/g, ''));
+    });
```
Comment (IvanGoncharov) [inline src/utilities/__tests__/mergeAST-test.js]: We don't use fixtures please unroll it here.
Use `dedent` to remove indentation from strings.
And defined `expectMerged` wrapper to not repeat `print(mergeAST(parse`

File: src/utilities/__tests__/mergeAST-test.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ *  LICENSE file in the root directory of this source tree.
+ *
+ * @flow strict
+ *
+ */
+
+import { expect } from 'chai';
+import { describe, it } from 'mocha';
+
+import { parse, print } from '../../index';
```
Comment (IvanGoncharov) [inline src/utilities/__tests__/mergeAST-test.js]: Use direct imports from appropriate files.

File: src/utilities/mergeAST.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+} from '../language/ast';
+
+/**
+ * Given a document AST, merge all fragment definitions into the
+ * provided operations using inline fragments.
+ */
+
+export function mergeAST(ast: DocumentNode): DocumentNode {
+  const fragmentDefinitions: { [name: string]: FragmentDefinitionNode } = {};
+  const inlinedFragments: { [name: string]: FragmentSpreadNode } = {};
```
Comment (IvanGoncharov) [inline src/utilities/mergeAST.js]: Please use `Object.create(null)` here and use `ObjMap` for types.

File: src/utilities/mergeAST.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  type DocumentNode,
+  type FragmentDefinitionNode,
+  type FragmentSpreadNode,
+} from '../language/ast';
+
+/**
+ * Given a document AST, merge all fragment definitions into the
+ * provided operations using inline fragments.
+ */
+
```
Comment (IvanGoncharov) [inline src/utilities/mergeAST.js]: Remove empty line

File: src/utilities/mergeAST.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    FragmentDefinition(node) {
+      // Collect the existing FragmentDefinition nodes.
+      fragmentDefinitions[node.name.value] = node;
+    },
+  });
+
+  return visit(ast, {
+    FragmentSpread(node) {
+      // No repeats!
+      if (inlinedFragments[node.name.value]) {
```
Comment (IvanGoncharov) [inline src/utilities/mergeAST.js]: This is incorrect check, same fragment can be used in multiple places
```graphql
{
  person: {
    ...ProfileInfo
    friend: {
      ...ProfileInfo
    }
  }
}
```
Also please cover in tests.

**Comment by IvanGoncharov** [comment]
> I found a failing case that wasn't in the original unit tests:

I think I know the reason, I will fix this.

File: src/utilities/mergeAST.js (hunk 6 above)
Comment (acao) [reply src/utilities/mergeAST.js]: covered in unit tests now :)

File: src/jsutils/uniqueBy.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ * which is invoked for each element in array to generate
+ * the criterion by which uniqueness is computed.
+ *
+ * Simiar to _.uniqBy from lodash.
+ */
+export function uniqueBy(
+  array: $ReadOnlyArray<any>,
+  iteratee: (item: any) => any,
+) {
+  const FilteredMap = new Map();
```
Comment (langpavel) [inline src/jsutils/uniqueBy.js]: Why not `Set`?

**Comment by acao** [comment]
@langpavel good questions!

File: src/jsutils/uniqueBy.js (hunk 7 above)
Comment (acao) [reply src/jsutils/uniqueBy.js]: `Set` makes a lot more sense

File: src/utilities/inlineNamedFragments.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      fragmentDefinitions[definition.name.value] = definition;
+    }
+  }
+
+  return visit(documentAST, {
+    FragmentSpread(node) {
+      return {
+        ...fragmentDefinitions[node.name.value],
+        kind: 'InlineFragment',
+      };
```
Comment (IvanGoncharov) [inline src/utilities/inlineNamedFragments.js]: Please review type definitions before you use them or convert them into each other.
Fragment definition has way more fields than inline fragments.

File: src/utilities/inlineNamedFragments.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    },
+    SelectionSet(node) {
+      return {
+        ...node,
+        selections: uniqueBy(
+          node.selections,
+          selection => selection.name.value,
```
Comment (IvanGoncharov) [inline src/utilities/inlineNamedFragments.js]: You also removing duplicated fields, including fields with different aliases, args, etc.

File: src/jsutils/uniqueBy.js (hunk 7 above)
Comment (IvanGoncharov) [reply src/jsutils/uniqueBy.js]: `Map`, `Set` should be used only for non-string keys, `Object.create(null)` is perfectly fine for working with string keys.

File: src/utilities/inlineNamedFragments.js (hunk 8 above)
Comment (acao) [reply src/utilities/inlineNamedFragments.js]: for sure, adding more test cases, it can't be this simple

File: src/utilities/__tests__/mergeAST-test.js (hunk 2 above)
Comment (IvanGoncharov) [reply src/utilities/__tests__/mergeAST-test.js]: Please write it like this:
```js
function expectInlined(query) {
  return expect(print(inlineNamedFragments(parse(query))));
}

describe('inlineNamedFragments', () => {
  it('does not modify query with no fragments', () => {
    expectInlined(dedent`
      {
        id
      }
    `).to.equal(dedent`
      {
        id
      }
    `);
  });
```

**Comment by acao** [comment]
disabling until I've done more research

File: src/jsutils/uniqueBy.js (hunk 7 above)
Comment (langpavel) [reply src/jsutils/uniqueBy.js]: @IvanGoncharov `Set` is much better *primitive* for checking presence/uniqueness than `Object.create(null)`, both by math sense and by the fact — `Map`s and `Set`s are well tested and perfectly tuned for purpose. Misusing objects as hash maps, well, it's only relique. Polyfilling is not needed today.

**Comment by benjie** [comment]
I have a more thorough implementation of this here: https://github.com/graphql/graphiql/pull/1542

Let me know if you're interested into a PR to GraphQL.js

**Comment by IvanGoncharov** [comment]
@benjie I think it would make sense to test it in GraphiQL to refine API and implementation and latter move to `graphql-js`.
Also, the relay compiler supports this and several similar transformations so we can learn from them.
